# Predicting High-Quality White Wine Using Machine Learning

## Business Objective

The objective is to develop a classification model that predicts whether a white wine will be classified as high quality based on its physicochemical properties.

A high-quality wine is defined as one receiving a quality score of 7 or higher. Such a model could provide an initial data-driven quality assessment to support quality-control and product-screening processes in the wine and beverage industry.

## Research Questions

### Primary Research Question

Can a machine-learning classification model predict whether a white wine will receive a high-quality score based on its physicochemical properties?

### Secondary Research Questions

1. Which physicochemical properties have the strongest relationships with high-quality wine classification?

2. Which model—Logistic Regression or a tuned K-Nearest Neighbours Classifier—performs better at identifying high-quality white wines?

3. How well does the best model perform in terms of accuracy, precision, recall, F1-score, and ROC-AUC?

In [13]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

wine_df = pd.read_csv("winequality-white.csv", sep=";")

wine_df.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.0,0.27,0.36,20.7,0.045,45.0,170.0,1.0010,3.00,0.45,8.8,6
1,6.3,0.30,0.34,1.6,0.049,14.0,132.0,0.9940,3.30,0.49,9.5,6
2,8.1,0.28,0.40,6.9,0.050,30.0,97.0,0.9951,3.26,0.44,10.1,6
3,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6
4,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6


In [14]:
print("Number of rows:", wine_df.shape[0])
print("Number of columns:", wine_df.shape[1])

wine_df.info()

Number of rows: 4898
Number of columns: 12
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


## 2. Data Quality Checks

The dataset contains 4,898 white-wine observations and 12 columns.

All columns contain 4,898 non-null values, confirming that there are no missing values in the dataset. Therefore, no missing-value treatment, such as removing rows or filling values, is required.

The dataset will also be checked for duplicated rows before proceeding with the analysis and machine-learning model.

In [3]:
print("Number of duplicated rows:", wine_df.duplicated().sum())

Number of duplicated rows: 937


### Duplicate Treatment

The dataset contained 937 duplicated rows. These rows were removed to prevent identical observations from receiving excessive influence during model training and potentially appearing in both the training and test sets.

After removing duplicates, the cleaned dataset contains 3,961 unique observations.

In [4]:
wine_df = wine_df.drop_duplicates().reset_index(drop=True)

print("Number of rows after removing duplicates:", wine_df.shape[0])
print("Remaining duplicated rows:", wine_df.duplicated().sum())


Number of rows after removing duplicates: 3961
Remaining duplicated rows: 0


## 3. Target Definition and Binary Encoding

The original `quality` variable contains numerical scores ranging from 3 to 9. For this classification project, the scores are converted into two classes:

- `0` = standard-quality wine, with a quality score below 7
- `1` = high-quality wine, with a quality score of 7 or higher

The threshold of 7 was selected to distinguish higher-rated wines from the more common scores of 5 and 6. This is a project-defined classification threshold and should not be interpreted as a universal wine-industry standard.

The Boolean results are converted into binary integers using `.astype(int)`. This is rule-based binary target encoding rather than One-Hot Encoding or Label Encoding.